test/joseph/JosephAutoRebalance.t.sol

```diff
@@ -0,0 +1,1692 @@
+        uint256 miltonInitPool = 1000000 * 1e6;
+        uint256 stanleyInitBalance = 800000 * 1e18;
+        uint256 userPosition = 150000 * 1e6;
+
+        uint256 expectedMiltonBalance = 50750 * 1e6;
+        uint256 expectedStanleyBalance = stanleyInitBalance;
+
+        _executeRedeemUsdt(
+            autoRebalanceThreshold,
+            miltonStanleyRatio,
+            miltonInitPool,
+            stanleyInitBalance,
+            userPosition,
+            expectedStanleyBalance,
+            expectedMiltonBalance
+        );
+    }
+
+    function _clearAndSetupSmartContractsUsdt() private {
+        _miltonStorageUsdt = getMiltonStorage();
+
+        _ipTokenUsdt = getIpTokenUsdt(address(_usdt));
+
+        _stanleyUsdt = getMockCase0Stanley(address(_usdt));
+
+        _itfMiltonUsdt = getItfMiltonUsdt(
+            address(_usdt),
+            address(_iporOracle),
+            address(_miltonStorageUsdt),
+            address(_miltonSpreadModel),
+            address(_stanleyUsdt)
+        );
+
+        _itfJosephUsdt = getItfJosephUsdt(
+            address(_usdt),
+            address(_ipTokenUsdt),
+            address(_itfMiltonUsdt),
+            address(_miltonStorageUsdt),
+            address(_stanleyUsdt)
+        );
+
+        prepareIpTokenUsdt(_ipTokenUsdt, address(_itfJosephUsdt));
+        prepareJoseph(_itfJosephUsdt);
+        prepareMilton(_itfMiltonUsdt, address(_itfJosephUsdt), address(_stanleyUsdt));
+    }
+
+    function _clearAndSetupSmartContractsDai() private {
+        _miltonStorageDai = getMiltonStorage();
+        _ipTokenDai = getIpTokenDai(address(_dai));
+        _stanleyDai = getMockCase0Stanley(address(_dai));
+
+        _itfMiltonDai = getItfMiltonDai(
+            address(_dai),
+            address(_iporOracle),
+            address(_miltonStorageDai),
+            address(_miltonSpreadModel),
+            address(_stanleyDai)
+        );
+
+        _itfJosephDai = getItfJosephDai(
+            address(_dai),
+            address(_ipTokenDai),
+            address(_itfMiltonDai),
+            address(_miltonStorageDai),
+            address(_stanleyDai)
+        );
+
+        prepareIpTokenDai(_ipTokenDai, address(_itfJosephDai));
+        prepareJoseph(_itfJosephDai);
+        prepareMilton(_itfMiltonDai, address(_itfJosephDai), address(_stanleyDai));
+    }
+
+    function _executeProvideLiquidityUsdt(
+        uint256 autoRebalanceThreshold,
+        uint256 miltonStanleyRatio,
+        uint256 miltonInitPool,
+        uint256 stanleyInitBalance,
+        uint256 userPosition,
+        uint256 expectedStanleyBalance,
+        uint256 expectedMiltonBalance
+    ) internal {
+        _itfJosephUsdt.setAutoRebalanceThreshold(autoRebalanceThreshold);
+        _itfJosephUsdt.setMiltonStanleyBalanceRatio(miltonStanleyRatio);
+
+        deal(address(_usdt), address(_itfMiltonUsdt), miltonInitPool);
+
+        vm.startPrank(address(_itfMiltonUsdt));
```

Comment:
You can use `vm.prank(address(_itfJosephUsdt));` it affect only next interaction with BC

Reply:
Done.